kong/runloop/handler.lua

```diff
@@ -687,12 +687,20 @@ return {
   },
   header_filter = {
     before = function(ctx)
+      local var = ngx.var
+      local header = ngx.header
+
       if ctx.KONG_PROXIED then
         local now = get_now()
         -- time spent waiting for a response from upstream
         ctx.KONG_WAITING_TIME             = now - ctx.KONG_ACCESS_ENDED_AT
         ctx.KONG_HEADER_FILTER_STARTED_AT = now
       end
+
+      local upstream_header = constants.HEADERS.UPSTREAM_STATUS
+      if singletons.configuration.headers[upstream_header] then
+        header[upstream_header] = string.match(var.upstream_status, "(%S+)$")
```

Comment:
Unfortunately, `string.match` is part of the NYI (Not Yet Implemented) LuaJIT API (meaning it will not JIT compile, and thus prevent the creation of a trace in this chunk), which is a problem for such hot code paths. Introducing non-JITable code paths in the proxy path has given us performance regressions in the past, and we should be mindful about them.

Instead, you may use the PCRE API offered by `ngx.re`, or any other trick you may think of to extract this value in plain Lua (the former is probably healthier).

Reply:
If anyone is interested in viewing the [entire list](http://wiki.luajit.org/NYI#libraries_string-library) of NYI methods. As I was curious myself.

Reply:
As it is in our [CONTRUBITING.md](https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#writing-performant-code) file, we expect all contributors to have read this list already :)

Reply:
I was trying to be a little less forward about where I found it 😏

Reply:
Thanks for pointing this out. I never gave NYI LuaJIT methods much thought before. Will be more careful next time.

Reply:
@hbagdi also keep in mind that without the `"jo"` flag, `ngx.re.match` can be (much!) slower than `string.match`.

Reply:
@hishamhm Thank you for pointing this out!